### PR TITLE
Stop forwarding Google credentials into the container sandbox (Fixes #2946)

### LINKED ISSUE #2946: Container sandbox forwards GEMINI_API_KEY/GOOGLE_API_KEY and mounts gcloud credentials, bypassing the credential proxy
Container sandboxing forwards provider API keys into the container environment and mounts Google credential material read-only, bypassing the credential proxy that exists specifically to keep secrets on the host. Found while documenting the sandbox threat model for #2685.

## What crosses the boundary

`packages/cli/src/utils/sandbox-containers.ts`:

- `addContainerEnvVars` builds an `envMap` that forwards `GEMINI_API_KEY` and `GOOGLE_API_KEY` from the host process into the container with `--env`. These are raw long-lived keys, not short-lived tokens.
- `addContainerVolumeMounts` mounts `~/.config/gcloud` read-only into the container whenever it exists on the host, and mounts the file named by `GOOGLE_APPLICATION_CREDENTIALS` read-only while also setting that variable inside the container.

`~/.config/gcloud` can contain long-lived refresh credentials and service-account keys. `GOOGLE_APPLICATION_CREDENTIALS` typically points at a service-account JSON key, which can mint access tokens from anywhere.

## Why this is a defect and not a design choice

The credential proxy exists precisely so that stored secrets stay on the host: the container talks to a Unix socket and receives short-lived access tokens, key writes are refused, and the capability token is delivered over an inherited descriptor so it never appears in argv, the environment, or a readable mounted file. That is a careful design.

Forwarding `GEMINI_API_KEY` and `GOOGLE_API_KEY` in plain environment variables defeats it for those providers. Any process inside the container — including anything an LLM-generated command spawns — can read them from `/proc/self/environ` or simply `env`. The same applies to the mounted gcloud directory and ADC file.

The likely origin is upstream Gemini/Vertex auth plumbing that predates the proxy, carried forward without being reconciled against it. The result is that a user who chooses container sandboxing specifically to contain an untrusted model still exposes their Google credentials to it.

## Suggested resolution

Route these through the credential proxy like every other provider credential, so the container receives short-lived tokens rather than raw keys and files:

1. Stop forwarding `GEMINI_API_KEY` and `GOOGLE_API_KEY` via `--env`. Resolve them host-side through the proxy on demand.
2. Stop mounting `~/.config/gcloud` and the ADC file by default. If Vertex or ADC flows genuinely need host material, mediate them through the proxy the way OAuth already is.
3. If some path truly cannot be proxied, make it opt-in and loud rather than automatic and silent, so the user chooses to widen the boundary.

Custom mounts supplied through `LLXPRT_SANDBOX_MOUNTS` / `SANDBOX_MOUNTS` and profile `mounts` are a separate, legitimate case: the user asked for those explicitly.

## Acceptance criteria

- Neither `GEMINI_API_KEY` nor `GOOGLE_API_KEY` appears in the container environment by default.
- `~/.config/gcloud` and `GOOGLE_APPLICATION_CREDENTIALS` are not mounted into the container by default.
- Gemini and Vertex requests still work from inside a container sandbox, using proxy-mediated credentials.
- Behavioral tests assert the absence of those variables and mounts in the generated run arguments, and cover the proxied path continuing to authenticate.
- Seatbelt behavior is unchanged; it runs on the host with full keyring access by design and is out of scope here.

## Documentation impact

`docs/sandbox.md` currently describes these crossings as deliberate, because that is what the code does today. Once this is fixed the page must be updated to state that stored provider credentials do not cross into the container. Tracking that alongside this fix.

## adendum

I don't udnerstand what good the gcloud stuff is and I'm good just not passing it for now. We already ahve the abiltiy to pass api keys through /key save /key load etc. there is no reason these should ahve been passed. Gneerally you shouldn't deploy to gcloud throough your cli but through ci/cd.

### PR DESCRIPTION
**Pull request by acoliver**
Fixes #2946

## The problem

Container sandboxing existed to contain an untrusted model, but it handed that model the user's Google credentials on the way in.

`packages/cli/src/utils/sandbox-containers.ts` did three things that defeated the credential proxy:

- `addContainerEnvVars` copied `GEMINI_API_KEY` and `GOOGLE_API_KEY` from the host process into the container with `--env`. Raw, long-lived keys — readable by any process in the container via `env` or `/proc/self/environ`.
- `addContainerVolumeMounts` mounted `~/.config/gcloud` read-only whenever it existed. That directory can hold refresh credentials and service-account keys.
- The same function mounted the file named by `GOOGLE_APPLICATION_CREDENTIALS` read-only and re-exported the variable inside the container. That file is typically a service-account JSON key that can mint access tokens from anywhere.

The credential proxy is careful work: the container talks to a Unix socket, receives short-lived tokens, key writes are refused, and the capability token arrives over an inherited descriptor so it never appears in argv, the environment, or a readable mounted file. Forwarding raw keys in plain environment variables threw that away for these providers.

## What changed

**The three crossings are gone.** No `--env GEMINI_API_KEY`, no `--env GOOGLE_API_KEY`, no gcloud mount, no ADC mount, no `GOOGLE_APPLICATION_CREDENTIALS` re-export.

**Non-secret configuration still crosses.** `GOOGLE_CLOUD_PROJECT`, `GOOGLE_CLOUD_LOCATION`, `GOOGLE_GENAI_USE_VERTEXAI`, `GOOGLE_GENAI_USE_GCA`, `GEMINI_MODEL`, `TERM`, and `COLORTERM` are configuration, not credentials, and are untouched.

**Mounts the user asked for still work.** `LLXPRT_SANDBOX_MOUNTS`, `SANDBOX_MOUNTS`, and profile `mounts` are unchanged and covered by explicit regression tests. The issue called these out as a separate, legitimate case.

**Seatbelt is untouched**, as the issue specified.

## Making Gemini still work from inside the container

Removing the env forwarding is only half the job — the container has to be able to authenticate some other way. It does: keys saved on the host with `/key save` and referenced by `auth-key-name` resolve over the proxy socket.

Verifying that path turned up two places where named-key resolution still reached for the host keyring directly instead of the sanctioned factory:

- `BaseProvider` called `getProviderKeyStorage()` (direct) rather than `createProviderKeyStorage()`.
- `CompressionProfileResolver` did the same, even though `provider-key-storage.ts` explicitly prohibits consumer calls to it.

Inside a container that storage is unreachable, so an `auth-key-name` would simply fail. Both now use `createProviderKeyStorage()`, which returns proxy-backed storage when `LLXPRT_CREDENTIAL_SOCKET` is set and the identical direct singleton otherwise. Host behavior is unchanged, and callers can still inject their own storage.

Every other named-key path in the codebase — `runtime/keyResolution.ts`, `runtime/profileApplication.ts`, `ui/commands/keyCommand.ts`, `profile-application/loadBalancerProfile.ts` — was already using the factory. These were the last two stragglers.

## A misleading error message

A user with `GEMINI_API_KEY` exported in their shell, running a sandbox, now hits the Gemini auth failure — and the old message told them to *set `GEMINI_API_KEY`*, which is precisely the thing that no longer crosses. When a credential socket is present the message now explains that ambient Google credentials are intentionally unavailable inside a sandbox and points at `/key save` plus `--key-name`. Outside a sandbox the message is unchanged. The check uses the same truthy test as the credential factories, so an empty `LLXPRT_CREDENTIAL_SOCKET` gets the host message, matching where the credentials actually come from.

## What this PR does not claim

The issue's addendum accepted that gcloud/ADC material simply would not cross for now. That has a consequence worth stating plainly rather than glossing: **Vertex AI no longer authenticates from inside a container sandbox.** A proxy-resolved named key authenticates against the Gemini Developer API, not Vertex, because Vertex requires the `vertex-ai` auth mode which a named key does not establish. `docs/sandbox.md` says this directly, and #2959 tracks proxying Vertex properly.

Two further crossings surfaced while verifying this one. Neither is in this issue's scope, and both are documented honestly in `docs/sandbox.md` rather than hidden behind a blanket "credentials do not cross" claim that would have been false:

- The LLxprt global config directory is still bind-mounted read-write, and it holds `profiles/*.json` — which `ProfileManager` persists with raw inline `auth-key` values — plus a global `.env`. Tracked as #2957.
- A project-level `.env` can set `SANDBOX_FLAGS`, letting an untrusted repository inject arbitrary engine flags (including `--env GEMINI_API_KEY`) into the host launcher. Tracked as #2958.

Fixing either would mean building a sanitized config-view subsystem or a launcher-variable trust boundary — new subsystems, not deletions.

## Tests

All new and modified tests are Bun-native and registered in `scripts/bun-test-manifest.ts`. This mattered more than it sounds: `packages/providers` runs entirely off that manifest, so a new test file that is not listed is silently never executed.

`packages/cli/src/utils/sandbox-containers.test.ts` — six behavioral tests driving the real generated argument array:

- `GEMINI_API_KEY` and `GOOGLE_API_KEY` set to sentinel values on the host produce no matching `--env` entry, and the sentinel values appear nowhere in the args.
- The non-secret configuration variables are still emitted as genuine adjacent `--env` / `KEY=value` pairs (parsed as pairs, not merely present somewhere in the array).
- A real gcloud fixture directory produces no `--volume` operand referencing it.
- A real `GOOGLE_APPLICATION_CREDENTIALS` file produces no mount and no re-export.
- `LLXPRT_SANDBOX_MOUNTS` and legacy `SANDBOX_MOUNTS` still produce their `--volume` entries.

The file's `vi.mock('node:child_process')` automock crashed under Bun (the shim walks module exports and hits a private-field getter), so it is now an explicit factory mock supplying only `execSync`, `spawn`, and `exec`. Every pre-existing `#1456` assertion still runs unchanged, under both runners.

`packages/providers/src/__tests__/BaseProvider.proxyKeyStorage.test.ts` — a real `CredentialProxyServer` on a real Unix socket. A `BaseProvider` subclass with `auth-key-name` resolves the key served over that socket, asserted at the provider's execution boundary. A second test proves an injected `providerKeyStorage` still wins even with a socket set.

`packages/agents/src/core/CompressionProfileResolver.proxyKeyStorage.test.ts` — same real-socket approach for the compression path, plus the boundary case where the proxy has no such key and the resolver must report no auth rather than falling through.

`packages/providers/src/gemini/GeminiProvider.auth.test.ts` — the sandbox-aware message, the host message, and the empty-socket case.

The removal guards were confirmed to genuinely fail against the pre-change implementation, not merely pass against the new one.

## Verification

`npm run test` (all workspaces green), `npm run test:scripts`, `npm run lint`, `npm run lint:eslint-guard`, `npm run typecheck`, `npm run format`, `npm run build`, and the smoke test all pass. No lint suppressions, type suppressions, severity downgrades, or threshold increases were added.

Two `test:scripts` files initially failed with `[vitest-worker]: Timeout calling "onTaskUpdate"` under load; re-run in isolation they pass 57/57.

## Summary by CodeRabbit

* **Security**
  * Improved sandbox credential isolation by preventing API keys, application credentials, and local cloud configuration from being forwarded or mounted automatically.
  * Added guidance for authenticating through the credential proxy and documented remaining credential-handling limitations.

* **Bug Fixes**
  * Named provider keys now resolve correctly through the credential proxy.
  * Added clearer authentication errors when sandbox credentials are unavailable.

* **Tests**
  * Expanded coverage for credential-proxy authentication, sandbox isolation, and supported configuration mounts.

### CODE REVIEW

File: packages/agents/src/core/CompressionProfileResolver.proxyKeyStorage.test.ts [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+      } else {
+        process.env.LLXPRT_CREDENTIAL_SOCKET = priorSocketEnv;
+      }
+      resetFactorySingletons();
+      if (server !== undefined) {
+        try {
+          await server.stop();
+        } catch {
+          // already stopped
+        }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/agents/src/core/CompressionProfileResolver.proxyKeyStorage.test.ts]: Agreed, and fixed by deleting the try/catch rather than narrowing it. This project prefers fail-fast over defensive handling: each test starts the server exactly once and never stops it, so `stop()` throwing means a real teardown defect (an unreleased Unix socket that would break the next run), and the test should fail loudly. The bare `catch` was hiding exactly the class of failure the comment claimed was benign.

[the remark this answers is not in the document]
Comment (acoliver) [reply packages/agents/src/core/CompressionProfileResolver.proxyKeyStorage.test.ts]: Agreed, and fixed by deleting the try/catch rather than narrowing it. This project prefers fail-fast over defensive handling: each test starts the server exactly once and never stops it, so `stop()` throwing means a real teardown defect (an unreleased Unix socket that would break the next run), and the test should fail loudly. The bare `catch` was hiding exactly the class of failure the comment claimed was benign.

File: packages/providers/src/__tests__/BaseProvider.proxyKeyStorage.test.ts [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+    }
+    resetFactorySingletons();
+    if (server) {
+      try {
+        await server.stop();
+      } catch {
+        // already stopped
+      }
+      server = undefined;
+    }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/__tests__/BaseProvider.proxyKeyStorage.test.ts]: Agreed, and fixed by deleting the try/catch rather than narrowing it. This project prefers fail-fast over defensive handling: each test starts the server exactly once and never stops it, so `stop()` throwing means a real teardown defect (an unreleased Unix socket that would break the next run), and the test should fail loudly. The bare `catch` was hiding exactly the class of failure the comment claimed was benign.

File: packages/cli/src/utils/sandbox-containers.test.ts [hunk 3]
```diff
@@ -13,9 +13,27 @@ import { DebugLogger, FatalSandboxError } from '@vybestack/llxprt-code-core';
...
+// Vitest replaces the entire module namespace with the factory return.
+vi.mock('node:child_process', async (importOriginal) => {
+  const actual: typeof import('node:child_process') = await importOriginal();
+  return {
+    ...actual,
+    execSync: vi.fn(),
+    spawn: vi.fn(),
+    exec: vi.fn(),
+  };
+});
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/cli/src/utils/sandbox-containers.test.ts:38]: Correct catch, and I took the safer of the two suggested resolutions: removed `exec: vi.fn()` so the real `exec` from `importOriginal` survives. `sandbox-containers.ts` evaluates `promisify(exec)` at module scope, so a stubbed `exec` would have produced an `execAsync` that never settles — a hang waiting for the first test to reach that path. Only `execSync` and `spawn` are stubbed now, since those are the process-launching calls these tests must not actually perform, and the comment was rewritten to say exactly that instead of asserting a requirement the code then contradicted.

[the remark this answers is not in the document]
Comment (acoliver) [reply packages/cli/src/utils/sandbox-containers.test.ts:38]: Correct catch, and I took the safer of the two suggested resolutions: removed `exec: vi.fn()` so the real `exec` from `importOriginal` survives. `sandbox-containers.ts` evaluates `promisify(exec)` at module scope, so a stubbed `exec` would have produced an `execAsync` that never settles — a hang waiting for the first test to reach that path. Only `execSync` and `spawn` are stubbed now, since those are the process-launching calls these tests must not actually perform, and the comment was rewritten to say exactly that instead of asserting a requirement the code then contradicted.

File: packages/cli/src/utils/sandbox-containers.test.ts [hunk 4]
```diff
@@ -194,3 +212,147 @@ describe.sequential('#1456 container network policy', () => {
...
+
+  it('still forwards non-secret Vertex/Gemini configuration vars', () => {
+    process.env.GOOGLE_CLOUD_PROJECT = 'my-project';
+    process.env.GOOGLE_CLOUD_LOCATION = 'us-central1';
+    process.env.GOOGLE_GENAI_USE_VERTEXAI = 'true';
+    process.env.GOOGLE_GENAI_USE_GCA = 'true';
+    process.env.GEMINI_MODEL = 'gemini-2.5-pro';
+
+    const args: string[] = [];
+    addContainerEnvVars(args, ENV_CONFIG, 'test-container', [], '/workspace');
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/cli/src/utils/sandbox-containers.test.ts:278]: Good point — that gap was real. Added `drops only the secrets when secret and non-secret vars are set together`, which sets `GEMINI_API_KEY` and `GOOGLE_API_KEY` to sentinels alongside `GOOGLE_CLOUD_PROJECT`, `GOOGLE_CLOUD_LOCATION`, and `GOOGLE_GENAI_USE_VERTEXAI`, then asserts the non-secret pairs are still emitted, the two secret names are absent, and the sentinel values appear nowhere in the joined args. A blanket GEMINI_/GOOGLE_ prefix blocklist would now fail it.

File: packages/agents/src/core/CompressionProfileResolver.proxyKeyStorage.test.ts [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        server = undefined;
+      }
+      fs.rmSync(tmpDir, { recursive: true, force: true });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/agents/src/core/CompressionProfileResolver.proxyKeyStorage.test.ts:160]: Declining this one. `fs.rmSync` already passes `force: true`, and the suggested remediation is to wrap it in a try/catch — which is the defensive-swallow pattern this project explicitly avoids, and the same pattern flagged as a problem in the sibling comments on `server.stop()`. Adding a swallow here would hide exactly the EBUSY/EPERM leak the comment is worried about instead of surfacing it. The related concern about `server.stop()` masking socket-unlink failures is addressed: that try/catch is gone, so a failure to release the socket now fails the test rather than being absorbed and then re-encountered here.

File: packages/providers/src/gemini/GeminiProvider.auth.test.ts [hunk 6]
```diff
@@ -344,4 +346,60 @@ describe('GeminiProvider Authentication', () => {
...
+    const mockAuthResolver = {
+      resolveAuthentication: vi.fn().mockResolvedValue(null),
+    };
+    process.env.LLXPRT_CREDENTIAL_SOCKET = '/tmp/test-cred-socket.sock';
+
+    const provider = createProviderWithRuntimeSettings();
+    (provider as unknown as { authResolver: unknown }).authResolver =
+      mockAuthResolver;
+
+    const error = await provider.getAuthMode().catch((e: unknown) => e);
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/gemini/GeminiProvider.auth.test.ts:364]: Declining. The premise that these tests could observe each other's leaked state does not hold here: this file already deletes `LLXPRT_CREDENTIAL_SOCKET` in both `beforeEach` and `afterEach`, so the `beforeEach` delete makes each test independent of whether a prior test cleaned up. Test files also do not share a process — Vitest isolates per file, and the Bun manifest runner (`scripts/run_bun_tests.ts`) deliberately spawns one process per file — so the cross-file mutation described cannot reach these cases. The save/restore helper would be a refactor of an established file-wide pattern for no behavioral gain.

File: packages/cli/src/utils/sandbox-containers.test.ts [hunk 7]
```diff
@@ -13,9 +13,29 @@ import { DebugLogger, FatalSandboxError } from '@vybestack/llxprt-code-core';
...
+// everything else in the module graph keeps the real implementations —
+// Vitest replaces the entire module namespace with the factory return.
+//
+// Only execSync and spawn are stubbed: those are the process-launching
+// calls these tests must not actually perform. `exec` is deliberately left
+// real, because sandbox-containers.ts evaluates promisify(exec) at module
+// scope; a stub there would produce an execAsync that never settles and
+// would hang the first test to reach that path.
+vi.mock('node:child_process', async (importOriginal) => {
+  const actual: typeof import('node:child_process') = await importOriginal();
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/cli/src/utils/sandbox-containers.test.ts:32]: The general claim about Bun is right, but it does not apply here, so I'm declining this one.

`vi.mock` in this repository is not Bun's raw `mock.module`. `bunfig.toml` preloads `test-setup/augment-bun-vi.ts`, a compat shim that intercepts `vi.mock` specifically to make Vitest-style factories work under Bun. It was built to solve exactly the deadlock described:

- The shim's own comment at `test-setup/augment-bun-vi.ts:271` names the failure mode — "causing deadlocks when importOriginal" — as the thing it exists to prevent.
- `augment-bun-vi.ts:472` defines `const importOriginal = (): unknown => syncActual;` — it returns an already-resolved synchronous value, loaded via `require()` (which bypasses `mock.module` interception) and pre-cached before mock registration.
- The comment at `augment-bun-vi.ts:436-437` states this directly: "importOriginal returns a sync value via require() so `await importOriginal()` resolves immediately."

So `await importOriginal()` never suspends on a dynamic `import()`; there is nothing for the loader to deadlock on.

This is also the established convention for Bun-native tests in this package rather than something new to this PR. `packages/cli/src/utils/sandbox-ssh-agent-preflight.test.ts`, which is already on main and already in the Bun manifest, mocks the same module the same way:

    vi.mock('node:child_process', async (importOriginal) => ({
      ...(await importOriginal<typeof import('node:child_process')>()),
      spawn: vi.fn(),
    }));

Empirically, `bun test src/utils/sandbox-containers.test.ts` passes 18/18, the file passes through the manifest runner (`bun scripts/run_bun_tests.ts --workspace cli`), and the "Bun Native Test Compatibility" CI check is green on this head. A deadlock would hang the runner rather than pass.

Making the factory synchronous would also break the Vitest side: this file runs under both runners, since `packages/cli`'s primary `test` script is still `vitest run`, and Vitest's `importOriginal()` genuinely returns a promise that must be awaited.
